spring-rabbit/src/main/java/org/springframework/amqp/rabbit/support/PublisherCallbackChannelImpl.java

```diff
@@ -670,53 +668,46 @@ private synchronized void generateNacksForPendingAcks(String cause) {
 
 	/**
 	 * Add the listener and return the internal map of pending confirmations for that listener.
-	 * Callers <b>must</b> synchronize on this channel object when modifying the map.
-	 * This method will be changed in a future release to NOT expose the map.
 	 * @param listener the listener.
-	 * @return the internal map of pending confirmations.
-	 * TODO: do not expose the map externally; change the {@code RabbitTemplate#getUnconfirmed(long)}
-	 * functionality to delegate to a method here.
 	 */
 	@Override
-	public synchronized SortedMap<Long, PendingConfirm> addListener(Listener listener) {
+	public void addListener(Listener listener) {
 		Assert.notNull(listener, "Listener cannot be null");
 		if (this.listeners.size() == 0) {
 			this.delegate.addConfirmListener(this);
 			this.delegate.addReturnListener(this);
 		}
-		if (!this.listeners.values().contains(listener)){
-			this.listeners.put(listener.getUUID(), listener);
-			this.pendingConfirms.put(listener, Collections.synchronizedSortedMap(new TreeMap<Long, PendingConfirm>()));
+		if (this.listeners.putIfAbsent(listener.getUUID(), listener) == null) {
+			this.pendingConfirms.put(listener, new ConcurrentSkipListMap<Long, PendingConfirm>());
```

Comment:
I can't still understand (find) why they use 4 index levels https://en.wikipedia.org/wiki/Skip_list, but I'm fine with the fix. 
Merging...